### DISCUSSION TITLE
Validation doesn't work on serializable custom cast attributes because toArray is not called on models

### DISCUSSION THREAD

**Discussion by realprabs** [livewire/livewire#3078] [7. Bugs (archived)]
### Description
When the model has an attribute with a custom cast and I try to validate the model, using `$this->validate()` in component, the validation doesn't work because validator is checking against instance of an object instead of primitive data. This is because `$value->toArray()` is not called on eloquent models in `getDataForValidation`.

### Exact steps to reproduce
```php
class Restaurant extends Model {
    protected $casts = ['commission_percentage' => DecimalCast::class,];
}

class DecimalCast implements CastsAttributes, SerializesCastableAttributes {
    public function get($model, $key, $value, $attributes) {
         if (!$value) return null;
         if ($value instanceof BigDecimal) return $value;
         return BigDecimal::of($value);
    }
    public function set($model, $key, $value, $attributes) {
        if ($value instanceof BigDecimal) return $value->toFloat();
        return (float) $value;
    }
    public function serialize($model, string $key, $value, array $attributes) {
        return $this->set($model, $key, $value, $attributes);
    }
}

class RestaurantForm extends Component {
    public Restaurant $restaurant;
    protected $rules = [
        'restaurant.commission_percentage' => 'required|numeric|between:0,100',
    ];
}
```

### Stripped-down, copy-pastable code snippets
Adding ` || $value instanceof Model` to the condition responsible for determining whether to call `$value->toArray()` in  `ValidatesInput.getDataForValidation` makes it work. Alternatively, we can use `method_exists($value, 'toArray')` as the condition.

```php
protected function getDataForValidation($rules) {
    // Rest of the code
    return collect($properties)->map(function ($value) {
            if ($value instanceof Collection || $value instanceof EloquentCollection || $value instanceof Model) return $value->toArray();
            return $value;
    })->all();
}
```

**Context**
- Livewire version: 2.3.0
- Laravel version: 8.10.0
- Browser: Any

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by basepack** [comment]
REOPEN

**Comment by basepack** [comment]
Possible related to: https://github.com/livewire/livewire/issues/1841

**Comment by iksaku** [comment]
Yeah, I think it could also be related to the issue with Accesors/Mutators, however, it would be worth trying this...

[In the first paragraph of the Custom Casters Documentation on Array/JSON casing](https://laravel.com/docs/8.x/eloquent-mutators#array-json-serialization), it is stated that, in order for a Casted object to be serialized, you need to add `Arrayable` or `JsonSerializable` in order to properly serialize the casted object.

In deed, covering for the `serialize()` method is also needed, but would love to know if you can get your caster working with the `Arrayable` or `JsonSerializable` interfaces.
